app/src/main/kotlin/com/wire/android/ui/home/conversations/usecase/GetConversationMessagesFromSearchUseCase.kt

```diff
@@ -17,57 +17,72 @@
     private val observeMemberDetailsByIds: ObserveUserListByIdUseCase,
-    private val messageMapper: MessageMapper
+    private val messageMapper: MessageMapper,
+    private val dispatchers: DispatcherProvider
 ) {
 
     /**
      * This operation combines messages searched from a conversation and its respective user to UI
-     * @param searchQuery The search term used to define which messages will be returned.
+     * @param searchTerm The search term used to define which messages will be returned.
      * @param conversationId The conversation ID that it will look for messages in.
      * @return A [Either<CoreFailure, List<UIMessage>>] indicating the success of the operation.
      */
     suspend operator fun invoke(
         searchTerm: String,
-        conversationId: ConversationId
-    ): Either<CoreFailure, List<UIMessage>> =
+        conversationId: ConversationId,
+        lastReadIndex: Int
+    ): Flow<PagingData<UIMessage>> {
+        val pagingConfig = PagingConfig(
+            pageSize = PAGE_SIZE,
+            prefetchDistance = PREFETCH_DISTANCE,
+            initialLoadSize = INITIAL_LOAD_SIZE
+        )
+
         if (searchTerm.length >= MINIMUM_CHARACTERS_TO_SEARCH) {
-            getConversationMessagesFromSearch(
+            return getMessagesSearch(
                 searchQuery = searchTerm,
-                conversationId = conversationId
-            ).map { foundMessages ->
-                foundMessages.flatMap { messageItem ->
-                    observeMemberDetailsByIds(
-                        userIdList = messageMapper.memberIdList(
-                            messages = foundMessages
-                        )
-                    ).map { usersList ->
-                        messageMapper.toUIMessage(
-                            userList = usersList,
-                            message = messageItem
-                        )?.let { listOf(it) } ?: emptyList()
-                    }.first()
+                conversationId = conversationId,
+                pagingConfig = pagingConfig,
+                startingOffset = max(0, lastReadIndex - PREFETCH_DISTANCE)
+            ).map { pagingData ->
+                pagingData.flatMap { messageItem ->
+                    observeMemberDetailsByIds(messageMapper.memberIdList(listOf(messageItem)))
+                        .mapLatest { usersList ->
+                            messageMapper.toUIMessage(usersList, messageItem)?.let { listOf(it) }
+                                ?: emptyList()
+                        }.first()
                 }
-            }
+            }.flowOn(dispatchers.io())
         } else {
-            Either.Right(value = listOf())
+            return flowOf(PagingData.empty<UIMessage>()).flowOn(dispatchers.io())
         }
+    }
 
     private companion object {
         const val MINIMUM_CHARACTERS_TO_SEARCH = 1
+        const val PAGE_SIZE = 20
+        const val INITIAL_LOAD_SIZE = 50
```

Comment:
50 for searching can be too high
lets set it as low as we can just so we can minimize the initial load time if there aren't too many messages that fit the search

Reply:
Do you think a default 20 (the same as page size) should be fine?

Reply:
to start it should be fine we can adjust as we see fit later